### ISSUE TITLE
Async pool gives up on reconnects

### ISSUE THREAD

**Issue by MannerMan** [psycopg/psycopg#370]
Hello!

I have an asyncio application which uses psycopg and psycopg-pool. Its purpose is to report database state over HTTP to clients, things like 'database is up and healthy'.

I would like a behavior were the application can start even if postgres is down, and keep attempting to reconnect essentially forever. According to the documentation, this should be 'the default' if I have interpreted it correctly;

> If an attempt to create a connection fails, a new attempt will be made soon after, using an exponential backoff to increase the time between attempts, until a maximum of reconnect_timeout is reached. When that happens, the pool will call the reconnect_failed() function, if provided to the pool, _and just start a new connection attempt_

However, the behavior I'm seeing is that as soon as `reconnect_timeout` is reached, the pool does not attempt more reconnects, and every query/request fails after it reaches `timeout`.

I have cut down my application to a minimal example to demonstrate, which initializes a connection pool and executes a simple query every second. If starting this application while postgres is shutdown, psycopg-pool will hit `reconnect_timeout` after 5 sec, and all consecutive queries after this point will fail due to no connection being available within `timeout`. Even if postgres is started again after this point, the pool will never recover/reconnect.
```python
import asyncio
import psycopg
import psycopg_pool
from psycopg_pool import AsyncConnectionPool
from psycopg.rows import dict_row
import logging
import os

LOG_LEVEL = logging.getLevelName(os.environ.get("LOG_LEVEL", "INFO"))

logging.basicConfig(format='{asctime} {levelname:5} [{name:14}] {message}', style='{', level=LOG_LEVEL)
logger = logging.getLogger('main')

CONNECTION_INFO = {
    "HOST": os.getenv('PG_HOST', 'localhost'),
    "DB": os.getenv('PG_DB', 'postgres'),
    "USER": os.getenv('PG_USER', 'postgres'),
    "PASS": os.getenv('PG_PASS', ''),
    "PORT": os.getenv('PG_PORT', '5432'),
    "MAX_CONN": int(os.getenv('PG_MAX_CONN', '3')),
    "MAX_WAITING": int(os.getenv('PG_MAX_WAITING', '50')),
    "CONN_TIMEOUT": int(os.getenv('PG_CONN_TIMEOUT', '2')),
    "POOL_RECONN_TIMEOUT": int(os.getenv('PG_POOL_RECONN_TIMEOUT', '5'))
}

class Database():
    def __init__(self, pg_connection):
        self.my_pool = None
        self.pg = pg_connection
        self.logger = logging.getLogger('database')
        self.active_connections = {}
        self.last_activity_update_error = False

    async def open_pool(self):
        self.my_pool = AsyncConnectionPool(
            conninfo=f"host={self.pg['HOST']} "
                     f"port={self.pg['PORT']} "
                     f"user={self.pg['USER']} "
                     f"password={self.pg['PASS']} "
                     f"dbname={self.pg['DB']} "
                     f"connect_timeout={self.pg['CONN_TIMEOUT']} ",
            min_size=1,
            max_size=self.pg['MAX_CONN'],
            max_waiting=self.pg['MAX_WAITING'],
            # Max time a client is allowed to wait for a connection from the pool
            timeout=self.pg['CONN_TIMEOUT'],
            # Max time pool attempts to reconnect before giving up..
            reconnect_timeout=self.pg['POOL_RECONN_TIMEOUT'],
            reconnect_failed=self.pool_connect_failure
        )
        self.my_pool.open(wait=False)
        self.logger.info('Initiated connection pool to db')

    async def close_pool(self):
        self.logger.info('Closing pool')
        await self.my_pool.close()
        self.logger.info('Closed connection pool to db')
    
    def pool_connect_failure(self, pool):
        self.logger.warning(f"called pool_connect_failure")

    async def execute_query(self, sql, parameters={}, fetch='all', timeout=5):
        try:
            async with self.my_pool.connection() as conn:
                await conn.set_autocommit(False)
                async with conn.cursor(row_factory=dict_row) as cur:
                    return await cur.execute(sql, parameters)
        except (psycopg_pool.PoolTimeout) as error:
            self.logger.error(f'Could not execute query due to pool error: {error}')
        except (psycopg_pool.PoolClosed) as error:
            self.logger.error(f'Could not execute query due to pool closed: {error}')

database = Database(CONNECTION_INFO)

async def state_loop():
    while True:
        print('attempting to run query')
        await database.execute_query(sql='select 1')
        await asyncio.sleep(1)

async def start():
    await database.open_pool()
    return asyncio.create_task(state_loop())

loop = asyncio.new_event_loop()

try:
    future_state_loop = loop.run_until_complete(start())
    loop.run_until_complete(future_state_loop)
except KeyboardInterrupt:
    future_state_loop.cancel()
    loop.run_until_complete(database.close_pool())
    print('bye')
```

Is this the expected behavior? Should I use reconnect_failed callback and attempt to re-initialize the pool somehow? 

Thank you for providing psycopg, it is an excellent library :)

**Comment by dvarrazzo** [comment]
Hello!

I haven't checked why your code failed. At a glance, you seemed right: there seems to be no code to reconnect after a failure. So I wrote a test to verify that this is a bug:

```python
# added to tests/pool/test_pool_async.py

@pytest.mark.slow
async def test_reconnect_retry(proxy):
    logging.getLogger("psycopg.pool").setLevel(logging.INFO)
    proxy.stop()

    ev = asyncio.Event()

    def failed(pool):
        ev.set()

    async with pool.AsyncConnectionPool(
        proxy.client_dsn, min_size=1, reconnect_timeout=1.0, reconnect_failed=failed
    ) as p:
        await asyncio.wait_for(ev.wait(), 2.0)

        proxy.start()

        async with p.connection(timeout=2) as conn:
            await conn.execute("select 1")
```

And... this code passes. After it is possible to reconnect (`proxy.start()`), a connection is obtained successfully.

What triggers the reconnection is the connection attempt itself: looking at the logs of the test run:

```
$ pytest tests/pool/test_pool_async.py::test_reconnect_retry -s --log-cli-level=info
...
tests/pool/test_pool_async.py::test_reconnect_retry 
--------------------------------------------------- live log call ---------------------------------------------------
WARNING  psycopg.pool:pool_async.py:422 error connecting in 'pool-1': connection failed: Connection refused
	Is the server running on that host and accepting TCP/IP connections?
WARNING  psycopg.pool:pool_async.py:422 error connecting in 'pool-1': connection failed: Connection refused
	Is the server running on that host and accepting TCP/IP connections?
WARNING  psycopg.pool:pool_async.py:422 error connecting in 'pool-1': connection failed: Connection refused
	Is the server running on that host and accepting TCP/IP connections?
WARNING  psycopg.pool:pool_async.py:424 reconnection attempt in pool 'pool-1' failed after 1.0 sec
INFO     root:fix_proxy.py:83 starting proxy
INFO     root:fix_proxy.py:92 proxy started
INFO     root:fix_proxy.py:127 proxy listening
INFO     psycopg.pool:pool_async.py:99 connection requested from 'pool-1'
INFO     psycopg.pool:pool_async.py:160 growing pool 'pool-1' to 1
INFO     psycopg.pool:pool_async.py:440 adding new connection to the pool
INFO     psycopg.pool:pool_async.py:135 connection given by 'pool-1'
INFO     psycopg.pool:pool_async.py:167 returning connection to 'pool-1'
PASSED
------------------------------------------------- live log teardown -------------------------------------------------
INFO     root:fix_proxy.py:106 stopping proxy
INFO     root:fix_proxy.py:109 proxy stopped

================================================= 1 passed in 1.39s =================================================
```

what happens is that, after the reconnection attempt failed, the number of available/potential connections (`pool._nconns`) is decreased. When `connection()` is called, no connection is found available, but `pool._nconns` is found to be < `conn_max, so a new connection attempt is started.

Said that, your code doesn't work indeed. The reason seems `_growing` set to True so `_maybe_grow_pool()` doesn't do its job. Looking into it better...

**Comment by dvarrazzo** [comment]
It seems an error managing the _growing flag indeed. That flag is necessary to avoid a spike of request to starve other database operations and makes sure that the pool can grow only one connection at time. However, if the growing connection fails, the flag is not reset.

This change seems what necessary to restore the pool state. Does it work for you (it works on your example).

```diff
diff --git a/psycopg_pool/psycopg_pool/pool_async.py b/psycopg_pool/psycopg_pool/pool_async.py
index 864a5001..84484993 100644
--- a/psycopg_pool/psycopg_pool/pool_async.py
+++ b/psycopg_pool/psycopg_pool/pool_async.py
@@ -428,6 +428,10 @@ class AsyncConnectionPool(BasePool[AsyncConnection[Any]]):
                 )
                 async with self._lock:
                     self._nconns -= 1
+                    # If we have given up with a growing attempt, let's allow
+                    # a new one.
+                    if growing and self._growing:
+                        self._growing = False
                 self.reconnect_failed()
             else:
                 attempt.update_delay(now)
```

I would like to reproduce the issue in the test suite and will release a new psycopg_pool version soon.

Please note that, in your test, `self.my_pool.open(wait=False)` requires an `await`. However that operation is no-op because the pool is not initialized with `open=False` and I assume it's just a glitch in your repro.

**Comment by dvarrazzo** [comment]
I managed to mistreat the pool enough to reproduce the issue in the test suite: see commit above.

**Comment by dvarrazzo** [comment]
Another shortcoming related to this issue is that the grow mode was actually meant only to increase connections from `size_min` to `size_max` if clients were requesting extra connections, not really to replenish a pool after it had been disconnected. For this reason, a broken check meant that, with the patch above, only a single connection would have been added to the pool. Fixed in c7bec71f, now the pool is fully restored to `size_min` after a reconnection.

**Comment by MannerMan** [comment]
> It seems an error managing the `_growing` flag indeed. That flag is necessary to avoid a spike of request to starve other database operations and makes sure that the pool can grow only one connection at time. However, if the growing connection fails, the flag is not reset.
> 
> This change seems what necessary to restore the pool state. Does it work for you (it works on your example)?
> 
> ```diff
> diff --git a/psycopg_pool/psycopg_pool/pool_async.py b/psycopg_pool/psycopg_pool/pool_async.py
> index 864a5001..84484993 100644
> --- a/psycopg_pool/psycopg_pool/pool_async.py
> +++ b/psycopg_pool/psycopg_pool/pool_async.py
> @@ -428,6 +428,10 @@ class AsyncConnectionPool(BasePool[AsyncConnection[Any]]):
>                  )
>                  async with self._lock:
>                      self._nconns -= 1
> +                    # If we have given up with a growing attempt, let's allow
> +                    # a new one.
> +                    if growing and self._growing:
> +                        self._growing = False
>                  self.reconnect_failed()
>              else:
>                  attempt.update_delay(now)
> ```

Yup, this did the trick - the pool now reconnects automatically even after the first `reconnect_timeout` is reached! :tada: 

> 
> I would like to reproduce the issue in the test suite and will release a new psycopg_pool version soon.
> 
Wonderful!

> Please note that, in your test, `self.my_pool.open(wait=False)` requires an `await`. However that operation is no-op because the pool is not initialized with `open=False` and I assume it's just a glitch in your repro.

Yeah that was local error in the snippet, but thanks for pointing it out! 

> Another shortcoming related to this issue is that the grow mode was actually meant only to increase connections from `size_min` to `size_max` if clients were requesting extra connections, not really to replenish a pool after it had been disconnected. For this reason, a broken check meant that, with the patch above, only a single connection would have been added to the pool. Fixed in [c7bec71](https://github.com/psycopg/psycopg/commit/c7bec71f9b3471ee3cadaf77e3b11f040f204d52), now the pool is fully restored to `size_min` after a reconnection.

Excellent finding! Thank you for providing such quick feedback & fix :+1:  We make good use of both psycopg2 and psycopg3 at the company I work for :) Please let me know if I can be of any further assistance in testing or similar.

**Comment by dvarrazzo** [comment]
psycopg_pool package 3.1.2 released, with this fix. Please let us know if it's all ok.

> Excellent finding! Thank you for providing such quick feedback & fix +1 We make good use of both psycopg2 and psycopg3 at the company I work for :) Please let me know if I can be of any further assistance in testing or similar.

If your company used psycopg heavily, [sponsoring the project](https://github.com/sponsors/dvarrazzo/) would be really useful. Fixing a bug such as this one has taken almost a day of work :slightly_smiling_face: 

I'm also open to know more advanced use cases from "heavy users". For instance if a feature such as disconnection detection (#247) would be useful.
